SICP_book/chapter_2_4.metta

```diff
@@ -0,0 +1,708 @@
+        )
+        (let*
+        (
+            (() (put get-record division2 (lambda2 $name $record_file ($record_by_name $name $record_file $record_by_name))))
+            (() (put get-salary division2 (lambda1 $record ($salary $record))))
+        ))))
+
+; a)
+(= (get-record $emp_name $division $division_recordfile)
+   ((get get-record $division) $emp_name $division_recordfile))
+
+; b)
+(= (get-salary $division $record)
+    ((get get-salary $division) $record))
+
+; c)
+; Let division_recordfiles contain pairs ($division_name . $record_file). Output is the first entry of emp_name in any
+; record file.
+(= (find-employee-record $emp_name Nil)
+    (Error (find-employee-record $emp_name Nil) "No such employee in input record files"))
+(= (find-employee-record $emp_name (Cons ($division_name . $record_file) $tail))
+    (let $get-rec-res (get-record $emp_name $division_name $record_file)
+        (if (== $get-rec-res Nil)
+            (find-employee-record $emp_name $tail)
+            $get-rec-res)))
+; -----------------------End of Exercise 2.74---------------------------
+
+
+; Message passing style:
+
+(= (make-from-real-imag_message $x $y)
+  (let $dispatch (lambda1 $op
+    (if (== $op real-part)
+        $x
+        (if (== $op imag-part)
+            $y
+            (if (== $op magnitude)
+                (sqrt (+ (square $x) (square $y)))
+                (if (== $op angle)
+                    (atan $y $x)
+                    (Error (make-from-real-imag_message $x $y) "Unknown operation"))))))
+  $dispatch))
+
+(= (apply-generic_message $op $arg) ($arg $op))
+
+; Exercise 2.75.  Implement the constructor make-from-mag-ang in
+; message-passing style. This procedure should be analogous to the
+; make-from-real-imag procedure given.
+
+(= (make-from-mag-ang_message $m $a)
+  (let $dispatch (lambda1 $op
+    (if (== $op real-part)
+        (* $m (cos $a))
+        (if (== $op imag-part)
+            (* $m (sin $a))
+            (if (== $op magnitude)
+                $m
+                (if (== $op angle)
+                    $a
+                    (Error (make-from-mag-ang_message $m $a) "Unknown operation"))))))
+  $dispatch))
+
+; Some asserts for those message passing functions:
+!(bind! &m_z1 (make-from-real-imag_message 5 2))
+!(bind! &m_z2 (make-from-mag-ang_message 3 4))
+
+;!(assertEqual
+;    (apply-generic_message real-part &m_z1)
+;    5)
+
+;!(assertEqual
+;    (apply-generic_message real-part &m_z2)
+;    -1.960930862590836)
+
+;!(assertEqual
+;    (apply-generic_message imag-part &m_z1)
+;    2)
+
+;!(assertEqual
+;    (apply-generic_message imag-part &m_z2)
+;    -2.2704074859237844)
+
+;!(assertEqual
+;    (apply-generic_message magnitude &m_z1)
+;    5.385164807134504)
+
+;!(assertEqual
+;    (apply-generic_message magnitude &m_z2)
+;    3)
+
+;!(assertEqual
+;    (apply-generic_message angle &m_z1)
+;    0.3805063771123649)
+
+;!(assertEqual
+;    (apply-generic_message angle &m_z2)
+;    4)
```

Comment:
Not sure why some code is commented out? Does it mean asserts are not stable or it is a performance issue?

Reply:
My bad. They work fine. I've pushed commit without ';'